### DISCUSSION TITLE
Feature Request: dynamic table names

### DISCUSSION THREAD

**Discussion by msto** [duckdb/duckdb#14539] [Ideas]
Hi,

Would it be possible for DuckDB to support dynamic or parametrized table names?

As best as I can tell, the prepared statements syntax only permits dynamically specified values, not table or column names.

https://duckdb.org/docs/sql/query_syntax/prepared_statements.html
https://duckdb.org/docs/api/python/dbapi#prepared-statements

```py
import duckdb

con = duckdb.connect()
con.execute("CREATE TABLE items (item VARCHAR, value DECIMAL(10, 2), count INTEGER)")
con.execute("INSERT INTO items VALUES ('jeans', 20.0, 1), ('hammer', 42.2, 2)")

con.execute("SELECT * FROM items WHERE item = $1", ["jeans"]).fetchall()
# [('jeans', Decimal('20.00'), 1)]

con.execute("SELECT * FROM items WHERE $1 = 'jeans'", ["item"]).fetchall()
# []

con.execute("SELECT * FROM $1 WHERE item = 'jeans'", ["items"]).fetchall()
# ---------------------------------------------------------------------------
# ParserException                           Traceback (most recent call last)
# Cell In[8], line 1
# ----> 1 con.execute("SELECT * FROM $1 WHERE item = 'jeans'", ["items"]).fetchall()
# 
# ParserException: Parser Error: syntax error at or near "$1"
```

The specific use case I am trying to solve for is the copying of several tables into DuckDB. I'd like to do something like the following, but I'm running into the `ParserException` exemplified above:
```py
table_names: list[str] = ["table_1", "table_2"]

with duckdb.connect("my.duckdb") as conn:
    for table_name in table_names:
        df: pd.DataFrame = get_table_contents(table_name)
        conn.execute(
            query="CREATE TABLE {table_name} AS SELECT * FROM df", 
            parameters={"table_name": table_name},
        )
```

Having dynamic table names seemed like the most straightforward solution, but I'm open to other suggestions.

I've previously used `psycopg.sql` to generate queries with dynamic table and field names. (I could pull in its composition functions and pass the generated query to DuckDB, but I'd prefer to avoid introducing other dependencies.) Would it be possible to implement something similar in DuckDB?

https://www.psycopg.org/psycopg3/docs/api/sql.html

Thanks!

**Comment by Tishj** [comment]
Sounds like <https://duckdb.org/docs/guides/sql_features/query_and_query_table_functions.html> should help?

**Reply by msto** [reply]
Thanks, I hadn't seen that!

It looks like `query_table` works for `SELECT` statements but not `CREATE TABLE` statements. Is that a deliberate omission?

```py
con.execute("SELECT * FROM query_table($1) WHERE item = 'jeans'", ["items"]).fetchall()
# [('jeans', Decimal('20.00'), 1)]

con.execute("CREATE TABLE query_table($1) (item VARCHAR, value DECIMAL(10, 2), count INTEGER)", ["items_2"])
# ---------------------------------------------------------------------------
# ParserException                           Traceback (most recent call last)
# Cell In[4], line 1
# ----> 1 con.execute("CREATE TABLE query_table($1) (item VARCHAR, value DECIMAL(10, 2), count INTEGER)", ["items_2"])
# 
# ParserException: Parser Error: syntax error at or near "$1"
```

**Reply by msto** [reply]
(I'm guessing it might be because the corresponding table does not exist when `query_table()` is called, so `query_table()` can't create a reference to it?)

**Reply by msto** [reply]
@Tishj is there a recommended solution for parameterized `CREATE` statements?

**Reply by msto** [reply]
FWIW, `query_table()` also does not work with `INSERT INTO` after creating a table. 

I understand why it wouldn't work for a `CREATE TABLE` statement, but once the table exists, why can't it be used with `INSERT INTO`?

```py
con = duckdb.connect()
con.execute("CREATE TABLE items (item VARCHAR, value DECIMAL(10, 2), count INTEGER)")

con.execute("INSERT INTO query_table($1) VALUES ('jeans', 20.0, 1), ('hammer', 42.2, 2)", ["items"])
# ---------------------------------------------------------------------------
# ParserException                           Traceback (most recent call last)
# Cell In[24], line 1
# ----> 1 con.execute("INSERT INTO query_table($1) VALUES ('jeans', 20.0, 1), ('hammer', 42.2, 2)", ["items"])
# 
#  ParserException: Parser Error: syntax error at or near "$1"

con.execute("INSERT INTO items VALUES ('jeans', 20.0, 1), ('hammer', 42.2, 2)")
con.execute("SELECT * FROM query_table($1)", ["items"]).fetchall()
# [('jeans', Decimal('20.00'), 1), ('hammer', Decimal('42.20'), 2)]

**Reply by andrewshandy** [reply]
Hey! I am wondering is there any solution to this? I have been getting the same error when I am using `httpfs` extension to write data to S3. My query is something similar to `connection.execute("COPY table TO $1", [my_s3_path]`, and get same error: `Parser Error: syntax error at or near "$"`

**Comment by lambeolino** [comment]
I have a similar problem, did you find a solution in the meantime @msto ? :)

**Reply by msto** [reply]
Just the ugly one :) 

The code I ended up with looked something like this:
```py
def insert_tables(conn: DuckDBPyConnection, tables: dict[str, pd.DataFrame]) -> None:
    """Insert the fetched tables into the DuckDB database."""
    
    table1_df = tables["table1"]  # noqa: F841
    table2_df = tables["table2"]  # noqa: F841
    # and so on. 
    
    conn.execute("INSERT INTO table1 SELECT * FROM table1_df")
    conn.execute("INSERT INTO table2 SELECT * FROM table2_df")
    # and so on
```

This was for a one-time client request last year, and we only had about a dozen tables to migrate, so we just bit the bullet on writing it out by hand. I haven't had the opportunity to revisit or follow up with the duckdb team about a better solution.

**Reply by lambeolino** [reply]
Okay, I think I also have to go the ugly way, thanks for the fast response! :)
